### ISSUE TITLE
Signers.V4 Error if non English characters were used in the request

### ISSUE THREAD

**Issue by anasqadrei** [aws/aws-sdk-js#2776]
- [x] I've gone though [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/welcome.html) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/)
- [x] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js) for answers
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js/issues) and didn't find any solution
  
**Bug Description**
I'm using the same signing method described in [Signing HTTP Requests to Amazon Elasticsearch Service](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-request-signing.html). Signing works fine with documents that has fields in English but not with other languages.

I've tried adding UTF-8 in content type but it didn't help.

**Is the issue in the browser/Node.js?**
Node.js

**If on Node.js, are you running this on AWS Lambda?**
Yes on both local machine and Lambda

**Details of the Node.js version**
Local machine: v8.9.3
Lambda v8.10

**SDK version number**
v2.483.0

**To Reproduce**
```javascript
// If your credentials don't work, export them at the terminal using the following commands:
// export AWS_ACCESS_KEY_ID="your-access-key"
// export AWS_SECRET_ACCESS_KEY="your-secret-key"

var AWS = require('aws-sdk');

var region = ''; // e.g. us-west-1
var domain = ''; // e.g. search-domain.region.es.amazonaws.com
var index = 'node-test';
var type = '_doc';
var id = '1';
var json = {
  "title": "عربي",  // CAUSE OF THE ERROR HERE. Change it to English and the error will disappear
  "director": "Bennett Miller",
  "year": "2011"
}

indexDocument(json);

function indexDocument(document) {
  var endpoint = new AWS.Endpoint(domain);
  var request = new AWS.HttpRequest(endpoint, region);

  request.method = 'PUT';
  request.path += index + '/' + type + '/' + id;
  request.body = JSON.stringify(document);
  request.headers['host'] = domain;
  request.headers['Content-Type'] = 'application/json';
  // charset doesn't work too
  // request.headers['Content-Type'] = 'application/json; charset=utf-8';

  // Content-Length is only needed for DELETE requests that include a request
  // body, but including it for all requests doesn't seem to hurt anything.
  request.headers["Content-Length"] = request.body.length;

  var credentials = new AWS.EnvironmentCredentials('AWS');
  var signer = new AWS.Signers.V4(request, 'es');
  signer.addAuthorization(credentials, new Date());

  var client = new AWS.HttpClient();
  client.handleRequest(request, null, function(response) {
    console.log(response.statusCode + ' ' + response.statusMessage);
    var responseBody = '';
    response.on('data', function (chunk) {
      responseBody += chunk;
    });
    response.on('end', function (chunk) {
      console.log('Response body: ' + responseBody);
    });
  }, function(error) {
    console.log('Error: ' + error);
  });
}
```

**Error**
```
403 Forbidden
Response body: {
  "message": "The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details."
}
```

**Additional context**
I understand that Signers.V4 is not part of the public API. It's for you to fix it. Also, I'm annoyed that it's part of AWS Elasticsearch documentation even though it has been recommended somehwere else ([#1247 (comment)](https://github.com/aws/aws-sdk-js/issues/1247#issuecomment-264285589)) not to use it for production.

**Comment by trivikr** [comment]
@anasqadrei Are you able to pass non-English characters in Java/Python/Ruby?

**Comment by anasqadrei** [comment]
@trivikr Nope. Sorry I'm not experienced in those languages. I'm going for holidays for 5 weeks soon. I can't investigate any further :(

**Comment by dangreaves** [comment]
This happens for me too, specifically with the `“` character. Looks like the signer is not correctly building the signature from a body containing those values.

They have a [new signer](https://github.com/aws/aws-sdk-js-v3/tree/master/packages/signature-v4-node) in v3 but not keen to begin using that until it's stable.

**Comment by trivikr** [comment]
Probably related https://github.com/aws/aws-sdk-js/issues/2706#issuecomment-501125092

**Comment by ajredniwja** [comment]
@anasqadrei can you try using the fix provided in https://github.com/aws/aws-sdk-js/issues/2706#issuecomment-501125092 
Please reach out if that doesn't fix the issue.

**Comment by anasqadrei** [comment]
@trivikr  @ajredniwja 

The issue in [#2706 (comment)](https://github.com/aws/aws-sdk-js/issues/2706#issuecomment-501125092) is related to `request.pathname`. My pathname is fine. It's all in English characters. 

However, my `request.body` has Arabic characters and is the cause of the problem. I've tried playing around in `v4.js` file but no luck. I've tried encoding the Arabic text using `aws.util.uriEscapePath`. The error disappeared but the text inserted into elastic search was encoded and couldn't be indexed/searched properly.

**Comment by ajredniwja** [comment]
@anasqadrei,

The signers class is not a public which is not tested throughly. 
Can you try encoding the body and let us know if it fixes the issue?

**Comment by anasqadrei** [comment]
@ajredniwja Yes, I have tried encoding the body. the request went through and I did not get the signer error. However, and because I am using the signer for **Elastic Search** service, the data were inserted into Elastic Search was encoded too. 

What I did was something like instead of:
`"title": "عربي"`
I changed it to:
`"title": AWS.util.uriEscapePath("عربي")`
I did not get a signer error but the data inserted was:
`"title": "%D8%B9%D8%B1%D8%A8%D9%8A"`

All interactions with Elastic Search happens though HTTP only.  Body get inserted as it is. If I send an encoded text, it will be inseated as such. I don't want that. I want it to be inserted in a human langauge where it can be indexed and searched.
